src/Simplex/Messaging/Agent.hs

```diff
@@ -1735,339 +1787,424 @@ processSMPTransmission c@AgentClient {smpClients, subQ} (tSess@(_, srv, _), v, s
-          | extSndId < prevExtSndId = MsgError $ MsgBadId extSndId
-          | extSndId == prevExtSndId = MsgError MsgDuplicate -- ? deduplicate
-          | extSndId > prevExtSndId + 1 = MsgError $ MsgSkipped (prevExtSndId + 1) (extSndId - 1)
-          | internalPrevMsgHash /= receivedPrevMsgHash = MsgError MsgBadHash
-          | otherwise = MsgError MsgDuplicate -- this case is not possible
+          qDuplex :: String -> (Connection 'CDuplex -> m ()) -> m ()
+          qDuplex name a = case conn of
+            DuplexConnection {} -> a conn
+            _ -> qError $ name <> ": message must be sent to duplex connection"
+
+          newRatchetKey :: CR.E2ERatchetParams 'C.X448 -> Connection 'CDuplex -> m ()
+          newRatchetKey e2eOtherPartyParams@(CR.E2ERatchetParams e2eVersion k1Rcv k2Rcv) (DuplexConnection cData' rqs sqs) = do
+            let rkHash = C.sha256Hash $ C.pubKeyBytes k1Rcv <> C.pubKeyBytes k2Rcv
+            exists <- withStore' c $ \db -> checkProcessedRatchetKeyHashExists db connId rkHash
+            unless exists $ do
+              withStore' c $ \db -> addProcessedRatchetKeyHash db connId rkHash
+              AgentConfig {e2eEncryptVRange} <- asks config
+              unless (e2eVersion `isCompatible` e2eEncryptVRange) (throwError $ AGENT A_VERSION)
+              keys <- getSendRatchetKeys
+              notifyAgreed
+              initRatchet e2eEncryptVRange keys
+            where
+              getSendRatchetKeys :: m (C.PrivateKeyX448, C.PrivateKeyX448, C.PublicKeyX448, Maybe (m ()))
+              getSendRatchetKeys
+                | rss == RSStarted = do
+                  (pk1, pk2, k1, _) <- withStore c (`getRatchetX3dhKeys'` connId)
+                  pure (pk1, pk2, k1, Nothing)
+                | otherwise = do
+                  (pk1, pk2, e2eParams@(CR.E2ERatchetParams _ k1 _)) <- liftIO . CR.generateE2EParams $ version e2eOtherPartyParams
+                  pure (pk1, pk2, k1, Just . void $ enqueueRatchetKeyMsgs c cData' sqs e2eParams)
+              notifyAgreed :: m ()
+              notifyAgreed = do
+                let cData'' = cData' {ratchetSyncState = RSAgreed} :: ConnData
+                    conn' = DuplexConnection cData'' rqs sqs
+                notify . RSYNC RSAgreed $ connectionStats conn'
+              recreateRatchet :: CR.Ratchet 'C.X448 -> m ()
+              recreateRatchet rc = withStore' c $ \db -> do
+                setConnRatchetSync db connId RSAgreed
+                deleteRatchet db connId
+                createRatchet db connId rc
+              -- compare public keys `k1` in AgentRatchetKey messages sent by self and other party
+              -- to determine ratchet initilization ordering
+              initRatchet :: VersionRange -> (C.PrivateKeyX448, C.PrivateKeyX448, C.PublicKeyX448, Maybe (m ())) -> m ()
+              initRatchet e2eEncryptVRange (pk1, pk2, k1, sendKeys_)
+                | C.pubKeyBytes k1 <= C.pubKeyBytes k1Rcv = do
```

Comment:
I think we probably should compare hashes and not keys themselves.

Reply:
why it matters?